File: yujung/프로그래머스/가장먼노드.cpp
#include <string>
#include <vector>
#include<queue>
//n개의 노드가 있는 그래프가 잇습니다. 각
using namespace std;
vector<int> v[20001];
int cnt[20001];

int visited[20001];
void bfs(int n){
    
    queue<int> q;
    
    q.push(n);
    visited[n]=true;
    cnt[n]=0;
   while(!q.empty()){
       int a=q.front();
       q.pop();
       for(int i=0; i<v[a].size();i++){
           if(visited[v[a][i]]==true)continue;
           visited[v[a][i]]=true;
           cnt[v[a][i]]=cnt[a]+1;
           q.push(v[a][i]);
       }
      
       }
}
int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    for(int i=0; i<edge.size();i++){
        v[edge[i][0]].push_back(edge[i][1]);
        v[edge[i][1]].push_back(edge[i][0]);
    }
    bfs(1);
    
    int max_n=0;
    for(int i=1; i<=n;i++){
        if(max_n<cnt[i]){
            max_n=cnt[i];
        }
    }
    
    for(int i=1; i<=n;i++){
        if(max_n==cnt[i]){
            answer++;
        }
    }
    return answer;
}

File: seoyeon/프로그래머스/알고리즘 고득점 Kit/그래프/가장 먼 노드.py
#BFS
from collections import deque
def solution(n, edge):
    answer = 0
    
    edge_n = [[] for _ in range(n+1)]
    visited = [False for _ in range(n+1)]
    
    for e in edge:
        x,y = e
        edge_n[x].append(y)
        edge_n[y].append(x)
    
    Q = deque()
    Q.append([0,1]) #[간선길이,노드]
    visited[1]=True
    max_len = 0
    while Q:
        len, node = Q.popleft()
        if len>=max_len:
            if len==max_len:
                answer += 1
            else:
                max_len = len
                answer = 1
        for next in edge_n[node]:
            if visited[next]==False:
                visited[next]=True
                Q.append([len+1,next])
    
    return answer